components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractLocalApplicationAuthenticator.java

```diff
@@ -0,0 +1,289 @@
+     * @throws AuthenticationFailedException the exception in the authentication flow
+     */
+    protected void processWithNonSaaSApp(AuthenticationContext context) throws AuthenticationFailedException {
+
+        String userTenantDomain = context.getSubject().getTenantDomain();
+        String spTenantDomain = context.getTenantDomain();
+        if (!StringUtils.equals(userTenantDomain, spTenantDomain)) {
+            context.setProperty("UserTenantDomainMismatch", true);
+            throw new AuthenticationFailedException("Service Provider tenant domain must be " +
+                    "equal to user tenant domain for non-SaaS applications", context.getSubject());
+        }
+    }
+
+    /**
+     * To process the authentication failed flow
+     *
+     * @param request  the httpServletRequest
+     * @param response the httpServletResponse
+     * @param context  the authentication context
+     * @return authentication flow status
+     * @throws AuthenticationFailedException the exception in the authentication flow
+     */
+    protected AuthenticatorFlowStatus initiateAuthenticationFlow(HttpServletRequest request,
+                                                                 HttpServletResponse response,
+                                                                 AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        if (getName().equals(context.getProperty(FrameworkConstants.LAST_FAILED_AUTHENTICATOR))) {
+            context.setRetrying(true);
+        }
+        initiateAuthenticationRequest(request, response, context);
+        context.setCurrentAuthenticator(getName());
+        return AuthenticatorFlowStatus.INCOMPLETE;
+    }
+
+    /**
+     * To check whether the user's account is being already locked or not.
+     *
+     * @param response httpServletResponse
+     * @param context  the authentication context
+     * @return true or false
+     * @throws AuthenticationFailedException the exception in the authentication flow
+     */
+    protected boolean checkUserAccountStatus(HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        String retryPage = ConfigurationFacade.getInstance().getAuthenticationEndpointRetryURL();
+        String queryParams = context.getContextIdIncludedQueryParams();
+        String errorCode = getErrorCode();
+        if (StringUtils.isNotEmpty(errorCode) && errorCode.equals(UserCoreConstants.ErrorCode
+                .USER_IS_LOCKED)) {
+            context.setRetrying(true);
+            context.setCurrentAuthenticator(getName());
+            try {
+                String redirectUrl = response.encodeRedirectURL(retryPage + ("?" + queryParams)) +
+                        FrameworkConstants.STATUS_MSG + FrameworkConstants.ERROR_MSG +
+                        FrameworkConstants.STATUS + FrameworkConstants.ACCOUNT_LOCKED_MSG;
+                response.sendRedirect(redirectUrl);
```

Comment:
can you move these to the caller method? The method name doesn't imply that you are doing some redirects or actions here. It just says 'check'